### ISSUE TITLE
Deploying with new stemcell didn't re-upload a new compiled release

### ISSUE THREAD

**Issue by drnic** [cloudfoundry/bosh#1620]
```
$ bosh2 -e vsphere-prod -d bosh-lite-44 deploy bosh-lite-44.yml

Release 'bosh/261.3' already exists.
```

In `bosh-lite-44.yml` I had bumped the stemcell version, and bumped the `releases:` for bosh to include a new url + sha1 that contained the compiled packages for the new stemcell.

I was hoping that bosh would download the new release (it did have a different sha1 from the previously uploaded `bosh/261.3`). Possible in future?

**Comment by cppforlife** [comment]
@drnic releases and stemcells are considered to be unrelated. for compiled releases of course there is a relationship but it allows for floating version constraints. i cant think of any easy way how director would now such information when it's downloading release from a single url.

suggestion to include source in compiled releases, doesnt really make sense to me since you are using compiled releases in a first place to avoid compilation.

**Comment by drnic** [comment]
I'm ok with not supporting src package collocation fundamentally but I think we still need a way to re-upload releases that contain new compiled packages for a new stemcell major release. As it stands now, I think we will cause deployments to fail if we upgrade major stemcells but one or more releases don't also need upgrading (their version is same so no way to install new compiled nor source packages).

Perhaps the algorithm could be for upload-release: check the version and sha1 of the release; if the version is different then we have some new jobs or packages to upload. If the sha1 is different then we have new compiled packages (for new major stemcell) to upload.





On Tue, Mar 14, 2017 at 8:46 PM -0400, "Dmitriy Kalinin" <notifications@github.com> wrote:












@drnic releases and stemcells are considered to be unrelated. for compiled releases of course there is a relationship but it allows for floating version constraints. i cant think of any easy way how director would now such information when it's downloading release from a single url.


suggestion to include source in compiled releases, doesnt really make sense to me since you are using compiled releases in a first place to avoid compilation.

**Comment by cppforlife** [comment]
note to self: how would this work with git urls.

**Comment by voelzmo** [comment]
Any ideas here? Uploading the same version of a release for a new major stemcell version should be possible without human interaction (i.e. `--fix` or leaving out `--name` and `--version`), right?

What's your workaround here?

**Comment by dsabeti** [comment]
@cppforlife We just hit this issue today. 

We had capi-release v1.29 compiled against stemcell v3363.24. Then, the stemcell version got bumped to version 3421. We recompiled the release against the new stemcell -- capi-release v1.29 with stemcell v3421. When we try to deploy, the BOSH director gives an error:
```
...

Release 'capi/1.29.0' already exists.

...

Can't use release 'capi/1.29.0'. It references packages without source code and are not compiled against stemcell 'bosh-aws-xen-hvm-ubuntu-trusty-go_agent/3421':
```

One way to understand this problem is that it impacts the ability to continuously deliver. As valuable as it is for bosh to compile releases as part of its deploy workflow, the vast majority of users use one of two or three stemcells and so would much rather use compiled releases. Especially with larger deployments like CF, we've gotten a lot of feedback that people prefer compiled releases. If we manage to solve this problem and ship compiled releases for everything, deploying a full Cloud Foundry becomes much more reasonable. 

We could definitely settle for some kind of flag, like @voelzmo suggested, that tells the CLI and the Director to force an update of the release even if the version numbers are the same. (My bet is that everyone would start using this flag all of the time, but you could do some validation for that.)  Another thing I noticed is that the SHA1 for the release changed when we compiled against a new stemcell: maybe the Director could (should?) notice that the SHA1 is different and insist on uploading the release.

cc @staylor14 @anEXPer

**Comment by dpb587** [comment]
Could we add support in the YAML/CLI to explicitly state a release URL reference as `compiled`?

From CLI, it'd require a new version which knows to `inspect-release` to check if the compiled version already appears. I suppose a smarter director endpoint could be added? I could imagine a potential bug/unfortunate behavior if a delete of the compiled release partially fails, leaving around a subset of packages, but the fix would be to remove the option, or retry deleting the release - either way, the user should have seen an earlier failure.

```
bosh upload-release \
  --name=openvpn \
  --version=3.2.1 \
  --compiled=ubuntu-trusty=3421.9 \
  https://...
```

The `releases` section could look similarly. The field would have to be updated whenever it's recompile's which is a slight shame, but that's the way it works for `version` already, so not a new argument.

Perhaps `compiled-for` / `exported` as alternatives.

**Comment by dsabeti** [comment]
+1 to a field like `compiled-for` that users can add in their manifest. This would certainly solve the issue for us in cf-deployment.

**Comment by staylor14** [comment]
Moar +1'ing for `compiled-for`.

**Comment by dpb587-pivotal** [comment]
I believe this is fixed in the [bosh-cli/v2.0.30](https://github.com/cloudfoundry/bosh-cli/releases/tag/v2.0.30). Example...

```
releases:
- name: zookeeper
  version: 0.0.5
  url: https://...compiled-bits.tgz
  stemcell:
    os: ubuntu-trusty
    version: 3421
```
